clang/lib/CIR/Dialect/Transforms/LoweringPrepareAArch64CXXABI.cpp

```diff
@@ -0,0 +1,176 @@
+//====- LoweringPrepareArm64CXXABI.cpp - Arm64 ABI specific code --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides ARM64 C++ ABI specific code that is used during LLVMIR
+// lowering prepare.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../IR/MissingFeatures.h"
+#include "LoweringPrepareItaniumCXXABI.h"
+
+#include <assert.h>
+
+using cir::LoweringPrepareCXXABI;
+using cir::MissingFeatures;
+
+using namespace mlir;
+using namespace mlir::cir;
+
+namespace {
+class LoweringPrepareAArch64CXXABI : public LoweringPrepareItaniumCXXABI {
+public:
+  LoweringPrepareAArch64CXXABI(clang::TargetCXXABI::Kind k) : Kind(k) {}
+  mlir::Value lowerVAArg(CIRBaseBuilderTy &builder,
+                         mlir::cir::VAArgOp op) override;
+
+private:
+  clang::TargetCXXABI::Kind Kind;
+  mlir::Value lowerGenericAArch64VAArg(CIRBaseBuilderTy &builder,
+                                       mlir::cir::VAArgOp op);
+};
+} // namespace
+
+LoweringPrepareCXXABI *
+LoweringPrepareCXXABI::createAArch64ABI(clang::TargetCXXABI::Kind k) {
+  return new LoweringPrepareAArch64CXXABI(k);
+}
+
+mlir::Value LoweringPrepareAArch64CXXABI::lowerGenericAArch64VAArg(
+    CIRBaseBuilderTy &builder, mlir::cir::VAArgOp op) {
+  auto loc = op->getLoc();
+  auto valist = op->getOperand(0);
+  auto opResTy = op.getType();
+  // front end should not produce non-scalar type of VAArgOp
+  bool isSupportedType =
+      opResTy.isa<mlir::cir::IntType, mlir::cir::SingleType,
+                  mlir::cir::PointerType, mlir::cir::BoolType,
+                  mlir::cir::DoubleType>();
+  assert(isSupportedType);
+
+  assert(!MissingFeatures::handleBigEndian());
```

Comment:
There's more than one place in OG codegen that checks for big endian, put the assert closer to those locations and repeat it as many as you see in OG codegen, with a comment before saying, example:
```
// TODO: handle bigendian in face of IsHSA or not aggregate for ABI... <whatever the content is>
assert(!MissingFeatures::handleBigEndian());
```
This one would go in "Argument was in registers" section. If you use variables with similar names with what OG uses that also could help here.